src/neural/cuda/layers.cc

```diff
@@ -1225,6 +1189,408 @@ ResidualBlock<DataType>::~ResidualBlock() {
+                 num_outputs * batch, num_outputs, num_outputs * batch, NONE,
+                 stream);
+    }
+
+    // LN1: skip connection and layer normilization
+    // scratch2/scratch0 -> scratch3
+    LayerNorm(N * 64, embedding_op_size_, scratch3, scratch2, scratch0,
+              enc.ln1_gammas, enc.ln1_betas, 1e-6, stream);
+
+    // #FFN dense 1, scratch3 -> scratch1
+    const int encoder_dff = enc.ffn_dense1_size_;
+    {
+      const int num_inputs = embedding_op_size_;
+      const int num_outputs = encoder_dff;
+      const int batch = N * 64;
+      cublasXgemm(cublas, CUBLAS_OP_T, CUBLAS_OP_N, num_outputs, batch,
+                  num_inputs, 1.0f, (const DataType*)enc.ffn_dense1_w,
+                  num_inputs, scratch3, num_inputs, 0.0f, scratch1,
+                  num_outputs);
+      addVectors(scratch1, (DataType*)enc.ffn_dense1_b, scratch1,
+                 num_outputs * batch, num_outputs, num_outputs * batch, SELU,
+                 stream);
+    }
+
+    // #FFN dense 2, scratch1 -> scratch2
+    {
+      const int num_inputs = encoder_dff;
+      const int num_outputs = embedding_op_size_;
+      const int batch = N * 64;
+      cublasXgemm(cublas, CUBLAS_OP_T, CUBLAS_OP_N, num_outputs, batch,
+                  num_inputs, 1.0f, (const DataType*)enc.ffn_dense2_w,
+                  num_inputs, scratch1, num_inputs, 0.0f, scratch2,
+                  num_outputs);
+      addVectors(scratch2, (DataType*)enc.ffn_dense2_b, scratch2,
+                 num_outputs * batch, num_outputs, num_outputs * batch, NONE,
+                 stream);
+    }
+
+    // LN2: skip connection and layer normilization
+    // scratch2/scratch3 -> scratch0
+    LayerNorm(N * 64, embedding_op_size_, scratch0, scratch2, scratch3,
+              enc.ln2_gammas, enc.ln2_betas, 1e-6, stream);
+
+  }  // End of encoder blocks
+
+  // queries (policy/attention/wq) -> scratch 1
+  {
+    const int num_inputs = embedding_op_size_;
+    const int num_outputs = policy_d_model_;
+    const int batch = N * 64;
+    cublasXgemm(cublas, CUBLAS_OP_T, CUBLAS_OP_N, num_outputs, batch,
+                num_inputs, 1.0f, ip2_pol_w_, num_inputs, scratch0, num_inputs,
+                0.0f, scratch1, num_outputs);
+    addVectors(scratch1, ip2_pol_b_, scratch1, num_outputs * batch, num_outputs,
+               num_outputs * batch, NONE, stream);
+  }
+
+  // keys (policy/attention/wk) -> scratch 2
+  {
+    const int num_inputs = embedding_op_size_;
+    const int num_outputs = policy_d_model_;
+    const int batch = N * 64;
+    cublasXgemm(cublas, CUBLAS_OP_T, CUBLAS_OP_N, num_outputs, batch,
+                num_inputs, 1.0f, ip3_pol_w_, num_inputs, scratch0, num_inputs,
+                0.0f, scratch2, num_outputs);
+    addVectors(scratch2, ip3_pol_b_, scratch2, num_outputs * batch, num_outputs,
+               num_outputs * batch, NONE, stream);
+  }
+
+  // dk = tf.math.sqrt(tf.cast(tf.shape(keys)[-1], self.model_dtype))  #
```

Comment:
I think auto format kind of messed up this comment a bit.

Reply:
fixed